components/form-builder/panel/Modal.tsx

```diff
@@ -0,0 +1,137 @@
+import React, { useEffect, useRef } from "react";
+import styled from "styled-components";
+import { useTranslation } from "next-i18next";
+import PropTypes from "prop-types";
+
+import { Button, FancyButton } from "./Button";
+import { Close } from "../icons/Close";
+
+const StyledDialog = styled.dialog`
+  width: 750px; /* TODO: fix this for mobile */
+  padding: 0;
+  background: transparent;
+  margin: 1.75rem auto;
+  background-color: #fff;
+  background-clip: padding-box;
+  border: 2px solid #000000;
+  box-shadow: 0px 4px 0px -1px #000000;
+  border-radius: 12px;
+  outline: 0;
+
+  .modal-content {
+    position: relative;
+    display: flex;
+    flex-direction: column;
+    width: 100%;
+    pointer-events: auto;
+  }
+
+  .modal-header {
+    display: flex;
+    align-items: flex-start;
+    justify-content: space-between;
+    padding: 15px;
+    border-bottom: 1px solid #e9ecef;
+  }
+
+  .modal-body {
+    position: relative;
+    flex: 1 1 auto;
+    padding: 15px;
+  }
+
+  .modal-footer {
+    display: flex;
+    align-items: center;
+    justify-content: flex-end;
+    padding: 15px;
+    border-top: 1px solid #e9ecef;
+  }
+`;
+
+export const Modal = ({
+  title,
+  children,
+  isOpen,
+  onClose,
+}: {
+  title: string;
+  children: JSX.Element | string;
+  isOpen: boolean;
+  onClose: () => void;
+}) => {
+  const { t } = useTranslation("form-builder");
+  const modalContainer = useRef<HTMLDialogElement>(null);
+  const close = () => {
+    modalContainer.current?.close();
+    onClose();
+  };
+
+  // focus modal when opened
+  useEffect(() => {
+    if (isOpen && modalContainer.current) {
+      modalContainer.current.showModal();
+      modalContainer.current.focus();
+    } else {
+      close();
+    }
+
+    document.body.style.overflow = isOpen ? "hidden" : "unset";
+  }, [isOpen]);
+
+  // Trap focus in the modal
+  useEffect(() => {
+    const handleFocusIn = ({ target }: FocusEvent) => {
+      if (modalContainer.current && !modalContainer.current.contains(target as Node)) {
+        modalContainer.current.focus();
+      }
+    };
+
+    document.addEventListener("focusin", handleFocusIn);
+    return () => document.removeEventListener("focusin", handleFocusIn);
+  }, []);
+
+  // Close modal if "ESC" key is pressed
+  useEffect(() => {
+    const handleKeyDown = (e: KeyboardEvent) => {
+      if (e.key === "Escape") {
+        e.stopPropagation();
+        close();
+      }
+    };
+
+    document.addEventListener("keydown", handleKeyDown);
```

Comment:
Believe we need to do a .current check here for React 18 (dev strict mode) --- which mounts and re-mounts causing the code to run twice. 

![react-18](https://user-images.githubusercontent.com/62242/187541663-ecc5fb63-bc0f-4cb3-a101-83dbe43246b2.gif)

Reply:
The listener unregisters itself with this code: 

```
// This adds the hook
document.addEventListener("keydown", handleKeyDown);
// this removes it when the component is removed
return () => document.removeEventListener("keydown", handleKeyDown);
```